File: lib/rgfa/line/common/dynamic_fields.rb
module RGFA::Line::Common::DynamicFields

  # Methods are dynamically created for non-existing but valid tag names.
  # Methods for predefined tags and positional fields
  # are created dynamically for each subclass; methods for existing tags
  # are created on instance initialization.
  #
  # ---
  #  - (Object) <fieldname>(parse=true)
  # The parsed content of a field. See also #get.
  #
  # <b>Parameters:</b>
  #
  # <b>Returns:</b>
  # - (String, Hash, Array, Integer, Float) the parsed content of the field
  # - (nil) if the field does not exist, but is a valid tag field name
  #
  # ---
  #  - (Object) <fieldname>!(parse=true)
  # The parsed content of a field, raising an exception if not available.
  # See also #get!.
  #
  # <b>Returns:</b>
  # - (String, Hash, Array, Integer, Float) the parsed content of the field
  #
  # <b>Raises:</b>
  # - (RGFA::NotFoundError) if the field does not exist
  #
  # ---
  #
  #  - (self) <fieldname>=(value)
  # Sets the value of a positional field or tag,
  # or creates a new tag if the fieldname is
  # non-existing but a valid tag name. See also #set, #set_datatype.
  #
  # <b>Parameters:</b>
  # - +*value*+ (String|Hash|Array|Integer|Float) value to set
  #
  # ---
  #
  def method_missing(m, *args, &block)
    super if virtual?
    field_name, operation, state = split_method_name(m)
    if ((operation == :get or operation == :get!) and args.size > 1) or
       (operation == :set and args.size != 1)
      raise RGFA::ArgumentError, "Wrong number of arguments \n"+
        "(method: #{m}; args.size = #{args.size})"
    end
    case state
    when :invalid
      super
    when :existing
      case operation
      when :get
        if args[0] == false
          field_to_s(field_name)
        else
          get(field_name)
        end
      when :get!
        if args[0] == false
          field_to_s!(field_name)
        else
          get!(field_name)
        end
      when :set
        set_existing_field(field_name, args[0])
        return nil
      end
    when :valid
      case operation
      when :get
        return nil
      when :get!
        raise RGFA::NotFoundError,
          "No value defined for tag #{field_name}"
      when :set
        set(field_name, args[0])
        return nil
      end
    end
  end

  # Redefines respond_to? to correctly handle dynamical methods.
  # @see #method_missing
  def respond_to?(m, include_all=false)
    super || (split_method_name(m)[2] != :invalid)
  end

  private

  def split_method_name(m)
    if @data.has_key?(m)
      return m, :get, :existing
    else
      case m[-1]
      when "!"
        var = :get!
        m = m[0..-2].to_sym
      when "="
        var = :set
        m = m[0..-2].to_sym
      else
        var = :get
      end
      if @data.has_key?(m)
        state = :existing
      elsif self.class::PREDEFINED_TAGS.include?(m) or
          valid_custom_tagname?(m)
        state = :valid
      else
        state = :invalid
      end
      return m, var, state
    end
  end

  #
  # Define field methods for a single field
  #
  def define_field_methods(fieldname)
    define_singleton_method(fieldname) do
      get(fieldname)
    end
    define_singleton_method :"#{fieldname}!" do
      get!(fieldname)
    end
    define_singleton_method :"#{fieldname}=" do |value|
      set_existing_field(fieldname, value)
    end
  end

  def self.included(base)
    base.extend(ClassMethods)
    base.instance_eval do
      private_class_method :apply_definitions
      private_class_method :define_field_accessors
      private_class_method :define_field_aliases
      private_class_method :define_reference_getters
    end
  end

  module ClassMethods

    def apply_definitions
      define_field_accessors
      define_field_aliases
      define_reference_getters
    end

    def define_field_accessors
      (self::POSFIELDS +
       self::PREDEFINED_TAGS).each do |fieldname|
        define_method(fieldname) do
          get(fieldname)
        end
        define_method :"#{fieldname}!" do
          get!(fieldname)
        end
        define_method :"#{fieldname}=" do |value|
          set_existing_field(fieldname, value)
        end
      end
    end
    def define_field_aliases
      if !self::NAME_FIELD.nil? and !self::POSFIELDS.include?(:name)
        self::FIELD_ALIAS[:name] = self::NAME_FIELD
      end
      self::FIELD_ALIAS.each do |k,v|
        alias_method :"#{k}",  :"#{v}"
        alias_method :"#{k}!", :"#{v}!"
        alias_method :"#{k}=", :"#{v}="
      end
    end

    def define_reference_getters
      (self::DEPENDENT_LINES + self::OTHER_REFERENCES).each do |k|
        if !method_defined?(k)
          define_method(k) do
            @refs ||= {}
            @refs.fetch(k, []).clone.freeze
          end
        end
      end
      if !method_defined?(:all_references)
        define_method :all_references do
          refs.values.flatten
        end
      end
    end
  end

end

File: gfapy/line/common/dynamic_fields.py
import gfapy
from types import MethodType

class DynamicField:
  def __init__(self, get, set):
    self.get = get
    self.set = set

class DynamicFields:
  """
  Methods are dynamically created for non-existing but valid tag names.
  Methods for predefined tags and positional fields
  are created dynamically for each subclass; methods for existing tags
  are created on instance initialization.
  """

  def __getattribute__(self, name):
    try:
      attr = super().__getattribute__(name)
      if not isinstance(attr, DynamicField):
        return attr
      else:
        return attr.get(self)
    except AttributeError as err:
      return self._get_dynamic_field(name, err)

  def __setattr__(self, name, value):
    try:
      attr = super().__getattribute__(name)
      if not isinstance(attr, DynamicField):
        return super().__setattr__(name, value)
      else:
        attr.set(self, value)
    except AttributeError:
      return self._set_dynamic_field(name, value)

  def _get_dynamic_field(self, name, err):
    if self.virtual:
      raise err
    if name.startswith("try_get_"):
      name = name[8:]
      try_get = True
    else:
      try_get = False
    if name in self._data:
      return (lambda : self.try_get(name)) if try_get else self.get(name)
    if (name in self.__class__.PREDEFINED_TAGS or
        self._is_valid_custom_tagname(name)):
      if not try_get:
        return None
      else:
        raise gfapy.NotFoundError(
          "No value defined for tag {}".format(name))
    else:
      raise err

  def _set_dynamic_field(self, name, value):
    try:
      virtual = super().__getattribute__("_virtual")
      data = super().__getattribute__("_data")
      if virtual:
        super().__setattr__(name, value)
      if name in data:
        self._set_existing_field(name, value)
      if (name in self.__class__.PREDEFINED_TAGS or
            self._is_valid_custom_tagname(name)):
        self.set(name, value)
      else:
        super().__setattr__(name, value)
    except AttributeError:
      super().__setattr__(name, value)

  def _define_field_methods(self, fieldname):
    """Define field methods for a single field"""
    def getter(self):
      return self.get(fieldname)
    def try_get(self):
      return self.try_get(fieldname)
    def setter(self, value):
      self._set_existing_field(fieldname, value)
    super().__setattr__(fieldname, DynamicField(getter, setter))
    super().__setattr__("try_get_" + fieldname, MethodType(try_get, self))

# XXX: class methods are missing
